### PR TITLE
Implement PioneerTestKit and PioneerEngineExecutionResults (#6)

### LINKED ISSUE #6: Replace Upstream Test Code With Dependency
#4 enabled tests to be written like in JUnit 5. Part of the implementation copy-pasted a bunch of classes from the project (see f613211 for reasons). As soon as JUnit publishes test artifacts (see junit-team/junit5#572) a proper dependency can be established and the class can be removed.

### PR DESCRIPTION
**Pull request by Bukama**
```
Up to now projects tests depends on copied internal JUnit Jupiter classes. In the past this was necessary but we always wanted to get rid of them. With the upgrade to JUnit 5.4.2 this is now possible, as the JUnit Jupiter team provides an public `EngineKit` as an API to those classes.

So with this PR, Pioneer gets its own `TestKit` and `EngineExectionResults` to wrap those of the JUnit Jupiter test engine as our tests always execute the `pioneer test engine`. This wrapper also provides more flexibility adding new functionality and/or easier access to things like the published `ReportEntries` of the executed test method.

PR: #218
closed #6 

ci skip-release
```
---

I hereby agree to the terms of the [JUnit Pioneer Contributor License Agreement](https://github.com/junit-pioneer/junit-pioneer/blob/master/CONTRIBUTING.md#junit-pioneer-contributor-license-agreement).

### CODE REVIEW

File: src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java [hunk 1]
```diff
@@ -83,26 +83,18 @@ void repeatedAnnotation_logEachKeyValuePairAsIndividualEntry() {
...
 	}
 
 	static class ReportEntriesTest {
 
+		private static AtomicInteger executionCount = new AtomicInteger(0);
+
 		@Test
 		@ReportEntry(key = "Crow2", value = "While I pondered weak and weary")
 		void explicitKey() {
+			System.out.println("HI THERE! " + executionCount.getAndIncrement());
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:97]: Why is this here?

File: src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java [hunk 2]
```diff
@@ -83,26 +83,18 @@ void repeatedAnnotation_logEachKeyValuePairAsIndividualEntry() {
...
-				.collect(toList());
-	}
-
 	private static Map.Entry<String, String> entryOf(String key, String value) {
 		return new AbstractMap.SimpleEntry<>(key, value);
 	}
 
 	static class ReportEntriesTest {
 
+		private static AtomicInteger executionCount = new AtomicInteger(0);
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:92]: Don't really see why this is here. Just for testing?

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+
+package org.junitpioneer.platform.testkit.engine;
+
+import static java.util.stream.Collectors.toList;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import org.junit.platform.engine.TestDescriptor;
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:19]: Test descriptor is 'unused' in the strictest sense because it is only mentioned in comments. I think we should use fully-qualified names in docs.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.platform.engine.TestExecutionResult;
+import org.junit.platform.engine.discovery.DiscoverySelectors;
+import org.junit.platform.testkit.engine.EngineExecutionResults;
+import org.junit.platform.testkit.engine.EngineTestKit;
+import org.junit.platform.testkit.engine.Events;
+
+/**
+ * Pioneer own class to handle {@link EngineExecutionResults} of the JUnit-Jupiter-Engine.
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:27]: 'Pioneers' to show possession. Sounds better to me.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+	 * Returns the published report entries of all tests.
+	 *
+	 * @return published report entries of all tests
+	 */
+	public List<Map<String, String>> getPublishedTestReportEntries() {
+		return executionResults
+				.tests()
+				.reportingEntryPublished()
+				.stream()
+				.map(executionEvent -> executionEvent.getPayload(org.junit.platform.engine.reporting.ReportEntry.class))
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:149]: You could import this (in which case the import is used over the `ReportEntry` in the same package).

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+				.filter(Optional::isPresent)
+				.map(Optional::get)
+				.map(org.junit.platform.engine.reporting.ReportEntry::getKeyValuePairs)
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:152]: You could import this.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+	 * Returns the execution results of the given method of a given test class.
+	 *
+	 * @param testClass Name of the test class
+	 * @param testMethodName Name of the test method (of the given class)
+	 * @return The execution results
+	 */
+	public static PioneerEngineExecutionResults getExecutionResults(Class<?> testClass, String testMethodName) {
+		PioneerEngineExecutionResults results = new PioneerEngineExecutionResults(testClass, testMethodName);
+
+		return results;
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:37]: You could inline this (i.e.: `return new Pioneer...`).

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * Returns the execution results of the given test class.
+	 *
+	 * @param testClass Name of the test class, the results should be returned
+	 * @return The execution results
+	 */
+	public static PioneerEngineExecutionResults getExecutionResults(Class<?> testClass) {
+		PioneerEngineExecutionResults results = new PioneerEngineExecutionResults(testClass);
+
+		return results;
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:24]: You could inline this (i.e.: `return new Pioneer...`).

File: src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java (hunk 1 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/jupiter/ReportEntryExtensionTest.java:97]: Nicolai added it for basic tests after creating the basic wrapper after updating so see if it works. And I just didn't think about it anymore. Will remove it

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java (hunk 7 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:37]: I know, but I normally don't do it, to have a better debug possibility and the costs for this variable are quite less in my opinion.

Also in this specific case (and the one below) the PioneerTestKit is a wrapper for future improving. We will then have to split it again.

**Comment by Bukama** [comment]
Those rebased and forced pushes on master really screw me up. Regardless if I rebase a branch on updated master or merge master into it. All changed files from master are shown in the PR....

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java (hunk 8 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerTestKit.java:24]: See above

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+import org.junit.platform.testkit.engine.EngineTestKit;
+import org.junit.platform.testkit.engine.Events;
+
+/**
+ * Pioneers own class to handle {@link org.junit.platform.testkit.engine.EngineExecutionResults} of the JUnit-Jupiter-Engine.
+ *
+ */
+public class PioneerEngineExecutionResults {
+
+	EngineExecutionResults executionResults;
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:31]: Is this package-private on purpose?

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+
+	public PioneerEngineExecutionResults(Class<?> testClass) {
+		executionResults = EngineTestKit
+				.engine("junit-jupiter")
+				.selectors(DiscoverySelectors.selectClass(testClass))
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:36]: Consider static importing `DiscoverySelectors` methods. Would that be more readable?

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+				.execute();
+	}
+
+	public PioneerEngineExecutionResults(Class<?> testClass, String testMethodName) {
+		executionResults = EngineTestKit
+				.engine("junit-jupiter")
+				.selectors(DiscoverySelectors.selectMethod(testClass, testMethodName))
```
Comment (Michael1993) [inline src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:43]: Consider a static import.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java (hunk 9 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:31]: Yes I always use package private instead of private to acces fields and methods from tests

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java (hunk 10 above)
Comment (Bukama) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:36]: Personally I like regular imports in such cases more as it's clear that this messages does not belong to the class. But if the majority wants to use static imports I can switch.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java (hunk 9 above)
Comment (Michael1993) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:31]: But in this case - do you access it? I'm not seeing the tests for this class.

File: src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java (hunk 10 above)
Comment (nicolaiparlog) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:37]: I like static imports in general, but I prefer them in situations where a somewhat universal API is used that is frequently seen. `DiscoverySelectors::selectClass` doesn't quite qualify, but I'm still ok, either way.

Comment (Bukama) [reply src/test/java/org/junitpioneer/platform/testkit/engine/PioneerEngineExecutionResults.java:37]: I would like to keep them. So I resolve this conversation.

**Comment by Bukama** [comment]
Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(

**Comment by Bukama** [comment]
> Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(

I mean: *NOW* I moved *ALL* extension tests to the new API and removed the superfluous classes.

But still two failing tests and I don't know f*** why 😢 If they would run, this would be mergeable.

**Comment by Bukama** [comment]
> > Moved all extension tests to the new API, but now two tests of the TempDirExtension fail because it seems that they are not started anymore and I can't figure out why :(
> 
> I mean: _NOW_ I moved _ALL_ extension tests to the new API and removed the superfluous classes.
> 
> But still two failing tests and I don't know f*** why 😢 If they would run, this would be mergeable.

NAILED them 🥳

**Comment by nicolaiparlog** [comment]
Great work, @Bukama! This was tough, but we made it. :tada: Let's hope we get a lot of green lights, so we can merge this.
